// @framer/danielkalman.satoshipay-ui-elements/code/canvas.tsx
// WARNING: this file is auto generated, any changes will be lost
import { createDesignComponent, CanvasStore } from "framer"
const canvas = CanvasStore.shared(); // CANVAS_DATA;
export const Back = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_iOBgquga2", {}, 178,28);
export const Big_Dropdown = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_c1fK65tb2", {}, 116,36);
export const Main_Menu = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_cPfAX87Pn", {}, 565,76);
export const Small_Button = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_nwjveHSGq", {}, 101,36);
export const TextField = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_eC3FkNfXT", {}, 432,62);
export const Transaction_Card = createDesignComponent<{parentSize?:{width:number|string,height:number|string},width?:number|string,height?:number|string}>(canvas, "id_vWw_aCSOz", {}, 684,42);

export const colors = Object.freeze({
    /** #35CEFF */
    "mainBlue": "var(--token-45e2f54e-993d-4d3e-a277-700c50b69aae, rgb(53, 206, 255))",
    /** #33C8FA */
    "hoverBlue": "var(--token-dd691f6a-6753-41ba-964d-132be3f59520, rgb(51, 200, 250))",
    /** #2BC1F3 */
    "darkerBlue": "var(--token-88e2bf77-8558-44b4-8bcf-662daf3a4a77, rgb(43, 193, 243))",
    /** #555555 */
    "darkGrey": "var(--token-9247dfd1-5faf-45eb-8c4f-de0d8cb20052, rgb(85, 85, 85))",
    /** #BBBBBB */
    "lightGrey": "var(--token-e0203bf4-a668-40e1-ac4a-4d7dbb026198, rgb(187, 187, 187))",
    /** rgba(34, 34, 34, 0) */
    "transparent": "var(--token-62cc917e-0d56-4c4a-a891-d00ea35d8713, rgba(34, 34, 34, 0))",
})
