// @framer/danielkalman.satoshipay-ui-elements/code/TextField.tsx
import * as React from "react"
import { Frame, addPropertyControls, ControlType } from "framer"
// @ts-ignore
import styled from "styled-components"
import { colors } from "./canvas"

// Open Preview: Command + P
// Learn more: https://framer.com/api

const StyledContainer = styled.div`
    width: 100%;
    height: 64px;
    `

const StyledInput = styled("input")`
    type: text;
    width: 100%;
    height: auto;
    position: relative;
    font-family: Lato;
    margin: 0;
    font-size: 18px;
    padding: 9px 0;
    border: 0;
    background-color: rgba(0,0,0,0);
    color: ${colors.darkGrey};

    &:focus {
        outline: 0;
    }
    `

const StyledLabel = styled.div`
    width: auto;
    padding: 0px;
    font-size: 12px;
    font-family: Rubik;
    font-weight: 500;
    letter-spacing: 1px;
    color: ${colors.lightGrey};
    text-transform: uppercase;
    position: relative;

    `

export function TextField(props) {
    const [inFocus, setInFocus] = React.useState(false)
    const [labelOn, setLabelOn] = React.useState(false)

    let inputCheck = ""

    return (
        <StyledContainer>
            <StyledLabel>{labelOn ? props.label : "\u00a0"}</StyledLabel>
            <StyledInput
                autocomplete="off"
                spellcheck="off"
                value={props.filled ? props.value : ""}
                placeholder={props.label}
                onFocus={() => {
                    setInFocus(true)
                    setLabelOn(true)
                }}
                onBlur={() => {
                    if (inputCheck.length < 1) {
                        setLabelOn(false)
                    } else {
                        setLabelOn(true)
                    }
                    setInFocus(false)
                }}
                onChange={event => {
                    inputCheck = event.target.value
                }}
            />
            <div
                style={{
                    width: "100%",
                    height: inFocus ? 2 : 1,
                    background: inFocus ? colors.mainBlue : colors.lightGrey,
                }}
            />
        </StyledContainer>
    )
}

TextField.defaultProps = {
    height: 55,
    width: 328,
    label: "Email",
    value: "",
    filled: false,
}

// Learn more: https://framer.com/api/property-controls/
addPropertyControls(TextField, {
    label: {
        title: "Label/Placeholder",
        type: ControlType.String,
        defaultValue: "Email",
    },
    value: {
        title: "Value",
        type: ControlType.String,
        defaultValue: "",
    },
    filled: {
        title: "Value",
        type: ControlType.Boolean,
        defaultValue: false,
    },
})
