// @framer/steveruizok.flow/code/_App.tsx
import { Override } from "framer"
// @ts-ignore
import { showNext, showPrevious, setRoot, showRoot, startStack } from "./index"

export function Flow(): Override {
    return {
        onChangePage: p => console.log(p),
        variants: {
            behind: {
                x: "-100%",
                transition: {
                    duration: 0,
                },
            },
            current: {
                x: "0%",
                transition: {
                    duration: 0,
                },
            },
            ahead: {
                x: "100%",
                transition: {
                    duration: 0,
                },
            },
        },
    }
}

export function SetRoot(): Override {
    return {
        onTap: () => setRoot(),
    }
}

export function BackToRoot(): Override {
    return {
        onTap: () => showRoot(),
    }
}

export function ShowD(): Override {
    return {
        onTap: () => showNext("pageD"),
    }
}

export function CustomTransition(): Override {
    return {
        onTap() {
            showNext("screen_c", {
                behind: {
                    x: "0%",
                    y: "0%",
                    filter: "brightness(50%)",
                    transition: {
                        ease: [0.23, -0.04, 0.31, 1.01],
                        duration: 0.36,
                    },
                },
                current: {
                    x: "0%",
                    y: "10%",
                    filter: "brightness(100%)",
                    transition: {
                        ease: [0.23, -0.04, 0.31, 1.01],
                        duration: 0.36,
                    },
                },
                ahead: {
                    x: "0%",
                    y: "100%",
                    filter: "brightness(100%)",
                    transition: {
                        ease: [0.23, -0.04, 0.31, 1.01],
                        duration: 0.28,
                    },
                },
            })
        },
    }
}
