// @framer/steveruizok.flow/code/App_Example.tsx
import { Override } from "framer"
// @ts-ignore
import { showNext, showPrevious } from "@framer/steveruizok.flow/code"

export function CustomTransition(): Override {
    return {
        onTap() {
            showNext("screen_c", {
                behind: {
                    x: "0%",
                    filter: "brightness(50%)",
                    transition: {
                        ease: [0.23, -0.04, 0.31, 1.01],
                        duration: 0.36,
                    },
                },
                current: {
                    x: "0%",
                    filter: "brightness(100%)",
                    transition: {
                        ease: [0.23, -0.04, 0.31, 1.01],
                        duration: 0.36,
                    },
                },
                ahead: {
                    x: "100%",
                    filter: "brightness(100%)",
                    transition: {
                        ease: [0.23, -0.04, 0.31, 1.01],
                        duration: 0.28,
                    },
                },
            })
        },
    }
}
