// @framer/steveruizok.flow/code/Marker.tsx
import * as React from "react"
import {
    Frame,
    Stack,
    FrameProps,
    addPropertyControls,
    ControlType,
    RenderTarget,
    useTransform,
    useMotionValue,
} from "framer"

type Props = Partial<FrameProps> & {
    name: string
}

export function Marker(props: Props) {
    const { name, ...rest } = props

    return RenderTarget.current() === RenderTarget.canvas ? (
        <Stack
            {...rest}
            height={44}
            color="#8855FF"
            background="rgba(136, 85, 255, 0.1)"
            border="1px solid #8855FF"
            style={{
                fontWeight: 600,
                fontSize: 14,
            }}
            borderRadius="22px 0px 0px 22px"
            direction="horizontal"
            alignment="center"
            distribution="start"
            paddingLeft={16}
        >
            <Frame
                size={16}
                border="2px solid #8855FF"
                background="none"
                borderRadius="100%"
            />
            <div style={{ paddingBottom: 1 }}>{name}</div>
        </Stack>
    ) : (
        <Frame background="none" size="100%" />
    )
}

Marker.defaultProps = {
    height: 44,
    width: 128,
    //
    name: "Home",
}

addPropertyControls(Marker, {
    name: { type: ControlType.String, title: "Name" },
})
